configuration.org
#+TITLE: Emacs Configuration File
#+AUTHOR: Timothée Gris

/Note : to link =init.el= to =configuration.org=, use the following :/

#+BEGIN_SRC emacs-lisp
;;(org-babel-load-file "~/.emacs.d/configuration.org")
#+END_SRC

* Packages initialization

#+BEGIN_SRC emacs-lisp
(package-initialize)

(require 'package)
(add-to-list 'package-archives '("melpa" . "http://melpa.org/packages/") t)
;;(add-to-list 'package-archives '("melpa" . "http://melpa.milkbox.net/packages/") t)

(package-initialize)

;; Update packages
(when (not package-archive-contents)
    (package-refresh-contents))
#+END_SRC

* Vim-like

** =evil-mode=

   Vim keybindings and integration in Emacs.
   
#+BEGIN_SRC emacs-lisp
(require 'evil)
(evil-mode 1)
#+END_SRC

** =powerline=

   =powerline= from Vim for Emacs.
  
#+BEGIN_SRC emacs-lisp
(require 'powerline)
#+END_SRC

** =powerline= theme

   Applies the default =powerline= theme.

#+BEGIN_SRC emacs-lisp
(powerline-default-theme)
#+END_SRC

* Basic Emacs modifications
  For a better Emacs experience.
** Enable =visual-line-mode=

   Words wrap at the end of a line.
   
#+BEGIN_SRC emacs-lisp
(global-visual-line-mode 1)
#+END_SRC

** Enable =linum-mode=

   Displays the line number on the left.

#+BEGIN_SRC emacs-lisp
(global-linum-mode t)
#+END_SRC

** Enable =minimap-mode=

   Displays a minimap on the right of the screen, /sublime-like/.

#+BEGIN_SRC emacs-lisp
;;(minimap-mode 1)

;; Move it to the right.

;;(setq minimap-window-location 'right)

;; Make it work with HTML and Org.

;;(add-to-list 'minimap-major-modes 'org-mode)
;;(add-to-list 'minimap-major-modes 'html-mode)

;;Change background color from old, distracting red to new, refreshing D3D3D3

;;(custom-set-faces
;;  '(minimap-active-region-background
;;    ((((background dark)) (:background "#2A2A2A222222"))
;;      (t (:background "#D3D3D3222222")))
;;    "Face for the active region in the minimap.
;;By default, this is only a different background color."
;;    :group 'minimap))
#+END_SRC

** Better window splitting

   Splits the window with the "next" buffer instead of the current buffer.

#+BEGIN_SRC emacs-lisp
(defun vsplit-last-buffer ()
  (interactive)
  (split-window-vertically)
  (other-window 1 nil)
  (switch-to-next-buffer)
  )
(defun hsplit-last-buffer ()
  (interactive)
  (split-window-horizontally)
  (other-window 1 nil)
  (switch-to-next-buffer)
  )
 
(global-set-key (kbd "C-x 2") 'vsplit-last-buffer)
(global-set-key (kbd "C-x 3") 'hsplit-last-buffer)
#+END_SRC

** Fullscreen on startup

   Emacs runs fullscreen when started.
   
#+BEGIN_SRC emacs-lisp
(defun fullscreen ()
       (interactive)
       (x-send-client-message nil 0 nil "_NET_WM_STATE" 32
                 '(2 "_NET_WM_STATE_FULLSCREEN" 0)))

;; comment line below to disable		 
;;(fullscreen)
#+END_SRC

** Emacs server

   Run emacs server on startup.

#+BEGIN_SRC emacs-lisp
(server-start)
#+END_SRC
   
** Notification flash

   Sets a visual on-screen flash instead of a notification beep.
   
#+BEGIN_SRC emacs-lisp
(setq visible-bell 1)
#+END_SRC
   
** Enable =pdf-tools=

   Better tool to view PDFs in Emacs (replaces DocView).

#+begin_src emacs-lisp
(pdf-tools-install)
#+end_src

* =org-mode= config
  
** =org-bullets=
   
   Better bullets instead of '*'.

#+BEGIN_SRC emacs-lisp
(require 'org-bullets)
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
#+END_SRC

** Export options
   
*** =ox-beamer=
    
    Export presentations in LaTeX !

#+BEGIN_SRC emacs-lisp
(require 'ox-beamer)

(require 'ox-latex)
(add-to-list 'org-latex-classes
             '("beamer"
               "\\documentclass\[presentation\]\{beamer\}"
               ("\\section\{%s\}" . "\\section*\{%s\}")
               ("\\subsection\{%s\}" . "\\subsection*\{%s\}")
               ("\\subsubsection\{%s\}" . "\\subsubsection*\{%s\}")))
#+END_SRC

*** =ox-twbs=

    Convert into HTML (TWBS).

#+BEGIN_SRC emacs-lisp
(require 'ox-beamer)
#+END_SRC

** =TODO= items states
   
   More choices for TODO items advencement.

#+BEGIN_SRC emacs-lisp
(setq org-todo-keywords
  '((sequence "TODO(t)" "RDV(r)" "WAIT(w@/!)" "|" "DONE(d!)" "CANCELED(c@)")))
#+END_SRC

** Syntax coloring in source code blocks

   Prettify source code blocks in =org-mode=.

#+BEGIN_SRC emacs-lisp
(setq org-src-fontify-natively t)
#+END_SRC
   
** Disable =linum-mode=

   Not only is =linum-mode= ugly and not very useful with org-mode, it takes an irreasonnable amount of resources.

#+begin_src emacs-lisp
(defun nolinum ()
  (global-linum-mode 0)
)
(add-hook 'org-mode-hook 'nolinum)
#+end_src

* Auto completion stuff

** =flycheck=

   Syntax checking on the fly.

#+begin_src emacs-lisp
;; always enabled
(global-flycheck-mode)
(add-hook 'after-init-hook #'global-flycheck-mode)
#+end_src

** General =company-mode= config  

   Triggers =company-mode= faster.

#+BEGIN_SRC emacs-lisp
(global-company-mode 1)
(setq company-dabbrev-downcase 0)
(setq company-idle-delay 0)
(setq company-minimum-prefix-length 2)
#+END_SRC

** C++

*** =irony-mode=

   Initializes =company-irony=.

#+BEGIN_SRC emacs-lisp
(eval-after-load 'company
  '(add-to-list 'company-backends 'company-irony))
#+END_SRC

   C++ auto-completion.

#+BEGIN_SRC emacs-lisp
(add-hook 'c++-mode-hook 'irony-mode)
(add-hook 'c-mode-hook 'irony-mode)
(add-hook 'objc-mode-hook 'irony-mode)

(add-hook 'irony-mode-hook 'irony-cdb-autosetup-compile-options)
#+END_SRC
   
** Python

*** =elpy=

   Here, we use Elpy : 
- Automatic indentation
- Syntax Highlighting
- Auto-Completion
- Syntax Checking
- Python REPL integration
- etc.

#+begin_src emacs-lisp
(elpy-enable)
(setq python-shell-interpreter "ipython"
      python-shell-interpreter-args "-i --simple-prompt")
#+end_src

*** =py-autopep8=

  and =py-autopep8= for Python standards (modifies whitespaces and indentations upon saving to follow those standards).

#+begin_src emacs-lisp
(require 'py-autopep8)
(add-hook 'elpy-mode-hook 'py-autopep8-enable-on-save)
#+end_src

* =tabbar-mode=

  Tabs in Emacs, with Dracula colors.

#+BEGIN_SRC emacs-lisp

;; Tabbar
(require 'tabbar)
;; Tabbar settings
(set-face-attribute
 'tabbar-default nil
 :background "#282a35"
 :foreground "#282a35"
 :box '(:line-width 1 :color "#282a35" :style nil))
(set-face-attribute
 'tabbar-unselected nil
 :background "#5c6370"
 :foreground "#abb2bf"
 :box '(:line-width 5 :color "#5c6370" :style nil))
(set-face-attribute
 'tabbar-selected nil
 :background "#282a35"
 :foreground "#abb2bf"
 :box '(:line-width 5 :color "#282a35" :style nil))
(set-face-attribute
 'tabbar-highlight nil
 :background "#abb2bf"
 :foreground "#282a35"
 :underline nil
 :box '(:line-width 5 :color "#abb2bf" :style nil))
(set-face-attribute
 'tabbar-button nil
 :box '(:line-width 1 :color "#282a35" :style nil))
(set-face-attribute
 'tabbar-separator nil
 :background "#282a35"
 :height 0.6)
(set-face-attribute
 'tabbar-modified nil
 :background "#5c6370"
 :foreground "#e06c75"
 :underline nil
 :box '(:line-width 5 :color "#5c6370" :style nil))
(set-face-attribute
 'tabbar-selected-modified nil
 :background "#282a35"
 :foreground "#e06c75"
 :underline nil
 :box '(:line-width 5 :color "#282a35" :style nil))

;; Changes padding of the tabs
;; we also need to set separators to avoid overlapping tabs by highlighted tabs
(custom-set-variables
 '(tabbar-separator (quote (0.5))))
;; adding spaces
(defun tabbar-buffer-tab-label (tab)
  "Return a label for TAB.
That is, a string used to represent it on the tab bar."
  (let ((label  (if tabbar--buffer-show-groups
                    (format "[%s]  " (tabbar-tab-tabset tab))
                  (format "%s  " (tabbar-tab-value tab)))))
    ;; Unless the tab bar auto scrolls to keep the selected tab
    ;; visible, shorten the tab label to keep as many tabs as possible
    ;; in the visible area of the tab bar.
    (if tabbar-auto-scroll-flag
        label
      (tabbar-shorten
       label (max 1 (/ (window-width)
                       (length (tabbar-view
(tabbar-current-tabset)))))))))

;; enable tabbars globally
(tabbar-mode 1)

#+END_SRC

* Keybindings
  
** =org-agenda=

  Set keybinding for =org-agenda= which were oddly missing

#+BEGIN_SRC emacs-lisp
(global-set-key (kbd "C-c a") 'org-agenda)
#+END_SRC
  
** Toggle menu bar
   
#+BEGIN_SRC emacs-lisp
(global-set-key [f9] 'toggle-menu-bar-mode-from-frame)
#+END_SRC
   
** =tabbar-mode=

   Navigate between tabs and groups.

#+BEGIN_SRC emacs-lisp
(global-set-key (kbd "C-S-p") 'tabbar-backward-group)
(global-set-key (kbd "C-S-n") 'tabbar-forward-group)
(global-set-key (kbd "C-<") 'tabbar-backward)
(global-set-key (kbd "C->") 'tabbar-forward) ;; tabbar.el, put all the buffers on the tabs.
#+END_SRC
   
** =crux=

   =crux-open-with= is used to open files, that Emacs doesn't support, with external programs.

#+begin_src emacs-lisp
(global-set-key (kbd "C-c o") 'crux-open-with)
#+end_src


eshell/history
ls
ls
ls
make Main.cpp 
g++ Main.cpp 
g++ Pokemon.cpp 
fish
cpp
cd Documents/epsi/c++/TP_Pokemon/Pokemon
ls
g++ Main.cpp
cd
ls
cd Downloads
ls
sudo mv menlo.ttf /usr/share/fonts
cd .emacs.d
cd
cd .emacs.d/
;s
l
ls
mv init.el init-nuit.el
ls
sh nuit.sh 
sh jour.sh 
ls
cd ..
ls
mkdir i3scrot
ls
mv i3scrot.sh i3scrot/
ls
rm i3lock_bar.sh 
rm i3lock_bar_veille.sh 
rm i3scrot_veille.sh 
mv veille.sh i3scrot/
cd i3scrot/
ls
cd ~/.config/i3status/
ls
vim config 
cd ..
ls
cd ProjetCPP/
ls
rm Main.cpp 
ls
rm ../Pokemon/Pokemon.cpp 
cp Pokemon.cpp ../Pokemon/
rm ../Pokemon/Pokemon.h 
cp Pokemon.h ../Pokemon/
ls
git add README.md 
git commit -m "Update on info about the project"
git push -u origin master
git add 
git add Main.cpp 
git commit -m "test"
git push -a origin timo
git push -u origin timo
git branch timo
git add Main.cpp 
git commit -m "test"
git push
git add Main.cpp 
git commit
git commit -m "test"
git push -u origin master
git branch timo
git checkout timo
git add Main.cpp 
git add Main.cpp 
git commit -m "test"
git push
git push -u origin timo
git add Main.cpp 
git commit -m "Test fin"
git push -u origin timo
cd ..
cd Documents/projets/cpp/Mastermind/
ls
./Main 
./Main 
./Main 
./Main 
./Main
./Main
./Main
./Main
ls
./Main
./Main
./Main
./Main
./Main
./Main
./Main
./Main
./Main
./Main
./Main
./Main
./Main
./Main 
./main
./Main


configuration.el
;;(org-babel-load-file "~/.emacs.d/configuration.org")

(package-initialize)

(require 'package)
(add-to-list 'package-archives '("melpa" . "http://melpa.org/packages/") t)
;;(add-to-list 'package-archives '("melpa" . "http://melpa.milkbox.net/packages/") t)

(package-initialize)

;; Update packages
(when (not package-archive-contents)
    (package-refresh-contents))

(require 'evil)
(evil-mode 1)

(require 'powerline)

(powerline-default-theme)

(global-visual-line-mode 1)

(global-linum-mode t)

;;(minimap-mode 1)

;; Move it to the right.

;;(setq minimap-window-location 'right)

;; Make it work with HTML and Org.

;;(add-to-list 'minimap-major-modes 'org-mode)
;;(add-to-list 'minimap-major-modes 'html-mode)

;;Change background color from old, distracting red to new, refreshing D3D3D3

;;(custom-set-faces
;;  '(minimap-active-region-background
;;    ((((background dark)) (:background "#2A2A2A222222"))
;;      (t (:background "#D3D3D3222222")))
;;    "Face for the active region in the minimap.
;;By default, this is only a different background color."
;;    :group 'minimap))

(defun vsplit-last-buffer ()
  (interactive)
  (split-window-vertically)
  (other-window 1 nil)
  (switch-to-next-buffer)
  )
(defun hsplit-last-buffer ()
  (interactive)
  (split-window-horizontally)
  (other-window 1 nil)
  (switch-to-next-buffer)
  )
 
(global-set-key (kbd "C-x 2") 'vsplit-last-buffer)
(global-set-key (kbd "C-x 3") 'hsplit-last-buffer)

(defun fullscreen ()
       (interactive)
       (x-send-client-message nil 0 nil "_NET_WM_STATE" 32
                 '(2 "_NET_WM_STATE_FULLSCREEN" 0)))

;; comment line below to disable		 
;;(fullscreen)

(server-start)

(setq visible-bell 1)

(pdf-tools-install)

(require 'org-bullets)
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))

(require 'ox-beamer)

(require 'ox-latex)
(add-to-list 'org-latex-classes
             '("beamer"
               "\\documentclass\[presentation\]\{beamer\}"
               ("\\section\{%s\}" . "\\section*\{%s\}")
               ("\\subsection\{%s\}" . "\\subsection*\{%s\}")
               ("\\subsubsection\{%s\}" . "\\subsubsection*\{%s\}")))

(require 'ox-beamer)

(setq org-todo-keywords
  '((sequence "TODO(t)" "RDV(r)" "WAIT(w@/!)" "|" "DONE(d!)" "CANCELED(c@)")))

(setq org-src-fontify-natively t)

(defun nolinum ()
  (global-linum-mode 0)
)
(add-hook 'org-mode-hook 'nolinum)

;; always enabled
(global-flycheck-mode)
(add-hook 'after-init-hook #'global-flycheck-mode)

(global-company-mode 1)
(setq company-dabbrev-downcase 0)
(setq company-idle-delay 0)
(setq company-minimum-prefix-length 2)

(eval-after-load 'company
  '(add-to-list 'company-backends 'company-irony))

(add-hook 'c++-mode-hook 'irony-mode)
(add-hook 'c-mode-hook 'irony-mode)
(add-hook 'objc-mode-hook 'irony-mode)

(add-hook 'irony-mode-hook 'irony-cdb-autosetup-compile-options)

(elpy-enable)
(setq python-shell-interpreter "ipython"
      python-shell-interpreter-args "-i --simple-prompt")

(require 'py-autopep8)
(add-hook 'elpy-mode-hook 'py-autopep8-enable-on-save)

;; Tabbar
(require 'tabbar)
;; Tabbar settings
(set-face-attribute
 'tabbar-default nil
 :background "#282a35"
 :foreground "#282a35"
 :box '(:line-width 1 :color "#282a35" :style nil))
(set-face-attribute
 'tabbar-unselected nil
 :background "#5c6370"
 :foreground "#abb2bf"
 :box '(:line-width 5 :color "#5c6370" :style nil))
(set-face-attribute
 'tabbar-selected nil
 :background "#282a35"
 :foreground "#abb2bf"
 :box '(:line-width 5 :color "#282a35" :style nil))
(set-face-attribute
 'tabbar-highlight nil
 :background "#abb2bf"
 :foreground "#282a35"
 :underline nil
 :box '(:line-width 5 :color "#abb2bf" :style nil))
(set-face-attribute
 'tabbar-button nil
 :box '(:line-width 1 :color "#282a35" :style nil))
(set-face-attribute
 'tabbar-separator nil
 :background "#282a35"
 :height 0.6)
(set-face-attribute
 'tabbar-modified nil
 :background "#5c6370"
 :foreground "#e06c75"
 :underline nil
 :box '(:line-width 5 :color "#5c6370" :style nil))
(set-face-attribute
 'tabbar-selected-modified nil
 :background "#282a35"
 :foreground "#e06c75"
 :underline nil
 :box '(:line-width 5 :color "#282a35" :style nil))

;; Changes padding of the tabs
;; we also need to set separators to avoid overlapping tabs by highlighted tabs
(custom-set-variables
 '(tabbar-separator (quote (0.5))))
;; adding spaces
(defun tabbar-buffer-tab-label (tab)
  "Return a label for TAB.
That is, a string used to represent it on the tab bar."
  (let ((label  (if tabbar--buffer-show-groups
                    (format "[%s]  " (tabbar-tab-tabset tab))
                  (format "%s  " (tabbar-tab-value tab)))))
    ;; Unless the tab bar auto scrolls to keep the selected tab
    ;; visible, shorten the tab label to keep as many tabs as possible
    ;; in the visible area of the tab bar.
    (if tabbar-auto-scroll-flag
        label
      (tabbar-shorten
       label (max 1 (/ (window-width)
                       (length (tabbar-view
(tabbar-current-tabset)))))))))

;; enable tabbars globally
(tabbar-mode 1)

(global-set-key (kbd "C-c a") 'org-agenda)

(global-set-key [f9] 'toggle-menu-bar-mode-from-frame)

(global-set-key (kbd "C-S-p") 'tabbar-backward-group)
(global-set-key (kbd "C-S-n") 'tabbar-forward-group)
(global-set-key (kbd "C-<") 'tabbar-backward)
(global-set-key (kbd "C->") 'tabbar-forward) ;; tabbar.el, put all the buffers on the tabs.

(global-set-key (kbd "C-c o") 'crux-open-with)
